src/core/fft_helpers.cpp

```diff
@@ -28,214 +29,251 @@
 #include <cmath>
 #include "lmms_constants.h"
 
-/* returns biggest value from abs_spectrum[spec_size] array
-
-   returns -1 on error
-*/
-float maximum(float *abs_spectrum, unsigned int spec_size)
-{
-	float maxi=0;
+/* Returns biggest value from abs_spectrum[spec_size] array.
+ *
+ *	returns -1 on error
+ *	returns maximum
+ */
+float maximum(float *abs_spectrum, unsigned int spec_size) {
+	float maxi = 0;
 	unsigned int i;
 
-	if ( abs_spectrum==NULL )
-		return -1;
+	if (abs_spectrum == NULL) {return -1;}
+	if (spec_size <= 0) {return -1;}
 
-	if (spec_size<=0)
-		return -1;
-
-	for ( i=0; i<spec_size; i++ )
-	{
-		if ( abs_spectrum[i]>maxi )
-			maxi=abs_spectrum[i];
+	for (i = 0; i < spec_size; i++) {
+		if (abs_spectrum[i] > maxi) {maxi = abs_spectrum[i];}
 	}
-
 	return maxi;
 }
 
+float maximum(std::vector<float> &abs_spectrum) {
+	return maximum(abs_spectrum.data(), abs_spectrum.size());
+}
 
-/* apply hanning or hamming window to channel
 
-	returns -1 on error */
-int hanming(float *timebuffer, int length, WINDOWS type)
-{
+/* Normalize the array of absolute magnitudes to a 0..1 range.
+ * Block size refers to FFT block size before any zero padding.
+ *
+ *	returns -1 on error
+ */
+int normalize(float *abs_spectrum, float *norm_spectrum, unsigned int bin_count, unsigned int block_size) {
 	int i;
-	float alpha;
 
-	if ( (timebuffer==NULL)||(length<=0) )
-		return -1;
+	if (abs_spectrum == NULL || norm_spectrum == NULL) {return -1;}
+	if (bin_count == 0 || block_size == 0) {return -1;}
+
+	for (i = 0; i < bin_count; i++) {
+		norm_spectrum[i] = abs_spectrum[i] / block_size;
+	}
+	return 0;
+}
+
+int normalize(std::vector<float> &abs_spectrum, std::vector<float> &norm_spectrum, unsigned int block_size) {
```

Comment:
Seems like a good idea in general, thanks. In this case, though, I'm not sure it would help -- it is just a wrapper / overloaded function that use vectors and the "real processing" is done over the raw underlying .data(), passed as a pointer. Making the pointer const would protect the pointer itself (which may be still a good idea), but it would not prevent any changes to the data, right?

Reply:
`const char* s` does not mean "declare s as a constant pointer to char" (to do that, you'd use `char* const s`), but it means "declare s as a pointer to const char". Similarly, `const std::vector<float>& v` does not mean "declare v as a constant reference to vector" (references can never be changed to reference to anything else, so they are always "constant"), but it means "declare v as a reference to const vector". Treating a vector const means you are only allowed to call its `const` functions, i.e. you can call `data() const`, but not `data()`. So passing the vector as const reference will guarantee that you will never modify the array behind the vector. Hope that clarifies, if not, keep asking.

Reply:
Thank you, I did not realize that `char* const s` is even a valid way how to use the const keyword. :)
I also did not know there is a constant version of data(), in this case it probably does make sense to use a constant vector here.

Reply:
Just a quick tip. If you want to know what a declaration is intended to do just read (or write) it from right to left. Examples:
* `char * const s`: `s` is a constant pointer to a (mutable) `char`.
* `char const * s`: `s` is a (mutable) pointer to a constant `char`.
* `char const * const s`: `s` is a constant pointer to a constant `char`.

Reply:
Reading it backwards seems like a nice idea. :) But in the 2 last cases, shouldn't the first const appear before the char?
I guess I'll have to review my operator precedence rules and stuff. I haven't used C/C++ a lot the last few years and some of the stuff apparently sublimated right out of my head. :))

Reply:
It would work both ways, i.e. the following two declarations describe the same thing:
* `char const * s`
* `const char * s`

However, if the `const` appears somewhere in the "middle" it's *not* the same if you swap the order: `char const * s` does not mean the same as `char * const s`.

It's one of the many confusing and ambiguous things in the standard and that's the reason why I always write the keywords to the right (because it always works in that case).

It also works for more complex declarations by the way, e.g.
* `char const * & r`: `r` is a reference to a pointer to immutable `chars`.

Reply:
@he29-net `const char c;` and `char const c;` is the same in C/C++. The const is just a type specifier (like volatile) which always [binds to the type keyword left or right to it](http://www.quut.com/c/ANSI-C-grammar-l-1998.html). When in doubt, you can enter any declaration in [cdecl](https://cdecl.org/). :smile:

Reply:
I added the const specifier where needed in fft_helpers.cpp in my local copy, but I noticed something unexpected: the internets say that the signature of a function should be the same whether or not it uses const on its parameters -- it's just an implementation detail. But when I only changed the definitions in .cpp file, it crashes at undefined function references at runtime. It's probably the same problem that killed the current github build.

When I add const to the declarations in header file as well, the problem goes away (as expected?). Any idea what causes this? I suspect the LMMS_EXPORT keyword may play some role here, and / or the fact that the plugin ends up being a separate .so library, so maybe some other rules apply to it..

Reply:
Not sure what internet link you've read. `const` is a type specifier; those must always be in definition and declaration, i.e. header and cpp file must be equal, concerning const-ness. This is different from storage class specifiers like `static` or `extern`. E.g. static member functions need the `static` keywords in the header, but not in the cpp file.

So if the error vanishes when you add const to the header aswell, that makes sense, as it must be there.

Reply:
I first got the idea from 
https://stackoverflow.com/questions/117293/use-of-const-for-function-parameters
and later
https://stackoverflow.com/questions/20659000/inconsistent-use-of-const-qualifier-between-declaration-and-definition
confirmed it by citing a paragraph from the standard:

> Parameter declarations that differ only in the presence or absence of const and/or volatile are equivalent. That is, the const and volatile type-specifiers for each parameter type are ignored when determining which function is being declared, defined, or called.

Based on this, I assume `const` does not have to be used in both definition and declaration, similar to default parameter values (where writing them twice is explicitly a bad idea).

Reply:
That's only if the *value* is passed const. The statement is invalid if the pointed-to memory is const.

```c++
void f(int); // 1
void f(const int); // 2
void g(char*); // 3
void g(const char*); // 4
void g(char* const); // 5
```
Can you guess which function declarations are equal?

Reply:
Well, based on what you say, 1 and 2 would be equal. And then perhaps 3 and 5, since the pointer is passed "as value" and the type it points to is same in both cases.
But what's the difference if the target memory is const? The standard further says:
> Only the const and volatile type-specifiers at the outermost level of the parameter type specification are ignored in this fashion

 -- is that what makes it different? I.e. the outermost level is the pointer and "const char" is therefore considered different from just "char"?

Reply:
> Well, based on what you say, 1 and 2 would be equal. And then perhaps 3 and 5, since the pointer is passed "as value" and the type it points to is same in both cases.

Exactly.

> -- is that what makes it different? I.e. the outermost level is the pointer and "const char" is therefore considered different from just "char"?

Yes. When you read a C declaration, you start "in the middle", and the outermost level in "pointer to const char" ist "const char".

Btw, it's very seldom that you declare a parameter (not the pointed to value) const. You can do this to make things easier to read in some cases. I've seen it more often that people write `const std::vector v` because they forgot the `&` :laughing:  than to clarify that the copied vector stays `const`.